x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,111 @@
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.ssl.SslUtil;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.http.HttpInfo;
+import org.elasticsearch.node.NodeService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.action.enrollment.CreateEnrollmentTokenAction;
+import org.elasticsearch.xpack.core.security.action.enrollment.CreateEnrollmentTokenRequest;
+import org.elasticsearch.xpack.core.security.action.enrollment.CreateEnrollmentTokenResponse;
+import org.elasticsearch.xpack.core.ssl.CertParsingUtils;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+import org.elasticsearch.xpack.security.authc.support.ApiKeyGenerator;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+import java.util.List;
+
+/**
+ * Transport action responsible for creating an enrollment token based on a request.
+ */
+
+public class TransportCreateEnrollmentTokenAction
+    extends HandledTransportAction<CreateEnrollmentTokenRequest, CreateEnrollmentTokenResponse> {
+
+    private final ApiKeyGenerator generator;
+    private final SecurityContext securityContext;
+    private final Environment environment;
+    private final NodeService nodeService;
+
+    @Inject
+    public TransportCreateEnrollmentTokenAction(TransportService transportService, ActionFilters actionFilters, ApiKeyService apiKeyService,
+                                                SecurityContext context, CompositeRolesStore rolesStore,
+                                                NamedXContentRegistry xContentRegistry, Environment environment, NodeService nodeService) {
+        super(CreateEnrollmentTokenAction.NAME, transportService, actionFilters, CreateEnrollmentTokenRequest::new);
+        this.generator = new ApiKeyGenerator(apiKeyService, rolesStore, xContentRegistry);
+        this.securityContext = context;
+        this.environment = environment;
+        this.nodeService = nodeService;
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateEnrollmentTokenRequest request,
+                             ActionListener<CreateEnrollmentTokenResponse> listener) {
+        createEnrolmentToken(request, listener);
+    }
+
+    private void createEnrolmentToken(CreateEnrollmentTokenRequest request, ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            generator.generateApiKeyForEnrollment(securityContext.getAuthentication(), request,
```

Comment:
I don't think we need to create specific methods in ApiKeyGenerator and ApiKeyService. The existing ones are meant to be generic enough to handle cases and we should probably add new methods for every time we need to create an API key. 

You can create a CreateApiKeyRequest object here based on the information of CreateEnrollmentTokenRequest and pass that to the existing methods.